libs/ojsonnet/Core_jsonnet.ml

```diff
@@ -161,11 +161,65 @@ and for_comp = tok (* 'for' *) * ident * tok (* 'in' *) * expr
 [@@deriving show { with_path = false }]
 
 (*****************************************************************************)
-(* Program *)
+(* Env *)
 (*****************************************************************************)
+type env = {
+  (* The spec uses a lambda-calculus inspired substitution model, but
+   * it is probably simpler and more efficient to use a classic
+   * environment where the locals are defined. Jsonnet uses lazy
+   * evaluation so we model this by allowing unevaluated expressions in environment below.
+   *)
+  locals : (local_id, local) Map_.t;
+  (* for call tracing *)
+  depth : int;
+}
 
-type program = expr [@@deriving show]
+and local_id = LSelf | LSuper | LId of string
+and local = Val of (value_ * env) | Expr of (expr * env)
+
+(*****************************************************************************)
+(* Values *)
+(*****************************************************************************)
+and value_ =
+  | Primitive of primitive
+  | ObjectVal of object_ AST_jsonnet.bracket
+  | Function of function_definition
+  | Array of (expr * env) array AST_jsonnet.bracket
```

Comment:
same here, it should be lazy_value array AST_jsonnet.bracket, otherwise
we would reevaluate each time array elements.